File: server-socket.c
// server.c — Servidor TCP simple que reconoce mensajes "i|texto"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>

#define PORT        18000
#define BACKLOG     10
#define BUF_SIZE    2048

int main(void) {
    int listen_fd = -1, conn_fd = -1;
    struct sockaddr_in addr, cliaddr;
    socklen_t clilen = sizeof(cliaddr);
    char buf[BUF_SIZE];

    // 1) Crear socket
    listen_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (listen_fd < 0) {
        perror("socket");
        return 1;
    }

    // Reusar puerto inmediatamente tras cerrar
    int yes = 1;
    if (setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes)) < 0) {
        perror("setsockopt SO_REUSEADDR");
        close(listen_fd);
        return 1;
    }

    // 2) Enlazar a 127.0.0.1:PORT (solo localhost)
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(PORT);
    if (inet_pton(AF_INET, "127.0.0.1", &addr.sin_addr) <= 0) {
        perror("inet_pton");
        close(listen_fd);
        return 1;
    }

    if (bind(listen_fd, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        perror("bind");
        close(listen_fd);
        return 1;
    }

    // 3) Escuchar
    if (listen(listen_fd, BACKLOG) < 0) {
        perror("listen");
        close(listen_fd);
        return 1;
    }

    printf("[Servidor] Escuchando en 127.0.0.1:%d ...\n", PORT);

    for (;;) {
        // 4) Aceptar conexión
        conn_fd = accept(listen_fd, (struct sockaddr*)&cliaddr, &clilen);
        if (conn_fd < 0) {
            perror("accept");
            continue; // intenta aceptar otro
        }

        char ipstr[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &cliaddr.sin_addr, ipstr, sizeof(ipstr));
        printf("[Servidor] Conexión de %s:%d\n", ipstr, ntohs(cliaddr.sin_port));

        // 5) Atender al cliente hasta que cierre
        for (;;) {
            ssize_t n = recv(conn_fd, buf, sizeof(buf) - 1, 0);
            if (n < 0) {
                perror("[Servidor] recv");
                break;
            } else if (n == 0) {
                printf("[Servidor] Cliente cerró la conexión\n");
                break;
            }
            buf[n] = '\0';

            printf("[Servidor] Recibido: %s\n", buf);

            char reply[BUF_SIZE];
            snprintf(reply, sizeof(reply), "OK %ld bytes recibidos | eco: %s", (long)n, buf);

            if (send(conn_fd, reply, strlen(reply), 0) < 0) {
                perror("[Servidor] send");
                break;
            }
        }

        close(conn_fd);
        conn_fd = -1;
    }

    close(listen_fd);
    return 0;
}

File: server-socket.py
# Explicación del programa:
# Este servidor TCP en Python utiliza hilos para poder atender a varios clientes de manera simultánea. Primero se crea
# un socket asociado al puerto 14000 y se lo configura para escuchar conexiones entrantes con una cola máxima de 5 en
# espera. Cuando un cliente se conecta, el servidor acepta la conexión y asigna un identificador incremental protegido
# por un candado (lock) para evitar condiciones de carrera entre múltiples hilos. Luego lanza un hilo independiente que
# ejecuta la función atender_cliente. Dentro de esta función se recibe un único mensaje enviado por el cliente, se imprime
# en la consola indicando de qué cliente proviene, se convierte el texto a mayúsculas y se devuelve al cliente como
# respuesta. Después se cierra la conexión y se informa en la salida estándar que la comunicación con ese cliente ha
# terminado. El servidor principal permanece en un bucle infinito aceptando nuevas conexiones y generando hilos para
# cada una. La entrada del programa son los mensajes que envían los clientes a través de sockets TCP y la salida consiste
# en dos partes: por un lado, la impresión en consola de los eventos de conexión, mensajes recibidos y cierre de
# conexiones; por otro, la respuesta que reciben los clientes, que es siempre el mismo mensaje convertido en mayúsculas.
# Un ejemplo de interacción sería que el cliente envíe "hola mundo", el servidor imprima en consola "[Cliente-1] Mensaje
# recibido: hola mundo" y devuelva al cliente la cadena "HOLA MUNDO". Después, al cerrarse la conexión, el servidor
# muestra "[Cliente-1] Conexión cerrada" y queda listo para aceptar nuevos clientes.

from socket import *
import threading

serverPort = 14000
serverSocket = socket(AF_INET, SOCK_STREAM)
serverSocket.bind(("", serverPort))
serverSocket.listen(5)   # permite hasta 5 conexiones en cola
print("The server is ready to receive")

contador = 0
lock = threading.Lock()

def atender_cliente(connectionSocket, addr, client_id):
    try:
        print(f"[Cliente-{client_id}] Conectado desde {addr}")
        sentence = connectionSocket.recv(1024).decode()
        print(f"[Cliente-{client_id}] Mensaje recibido: {sentence}")
        capitalizedSentence = sentence.upper()
        connectionSocket.send(capitalizedSentence.encode())
    except Exception as e:
        print(f"[Cliente-{client_id}] Error: {e}")
    finally:
        connectionSocket.close()
        print(f"[Cliente-{client_id}] Conexión cerrada")

while True:
    connectionSocket, addr = serverSocket.accept()
    with lock:
        contador += 1
        client_id = contador
    
    t = threading.Thread(target=atender_cliente, args=(connectionSocket, addr, client_id))
    t.start()
